### PR TITLE
The completely new and shiny Genetics Ark v2.0.0

### LINKED ISSUE #18: Handle when dx make_download_urls gives an error
Can give an error if I file has been moved that is in the json before it is replaced by new one with cron, need a way to handle what is returned

### LINKED ISSUE #19: Wide left hand padding
Left hand padding is irritating and needs removing without breaking everything

### LINKED ISSUE #22: Make development BAMs available through DNAnexus_to_igv
Would be good to have a project on DNAneuxs to put data in for scientists doing development work to be able to visualise bams themselves without us having to send links, this should be separate from the normal 002 projects and possibly run as a cron.hourly job to be updated throughout the day.

### LINKED ISSUE #24: DNAnexus - IGV and build 38
Does the search & IGV itself handle build 38 BAMs?
What will happen when we start processing in B38 & B37 concurrently (i.e. different services on different genome builds)?

### LINKED ISSUE #25: GUI accessibility tweak for IGV BAM urls
Please make it so that when you double click the URL (for BAM in DNAnexus), you highlight just the URL to allow easier copying.
Or even better, add a copy this URL button.

### LINKED ISSUE #32: Trust PC Firefox doesn't like Web IGV
Olivera wanted to like at X012421 on Genetics Ark and it was displaying an empty box instead of the IGV thingy.
It worked fine on Chrome

### PR DESCRIPTION
**Pull request by jethror1**
Mostly rewritten starting from a new base project and adding existing apps with required changes, changes include:

- updated to Django 3.1.1 from 1.11
- updated from python 2.7.? to 3.7+
- redesigned interface and navigation
- base of project: django_example -> ga_core
- all styling updated with bootstrap 4.0.0 and a main.css (in static/assets/css) - fixes #19 (+ no more super wide left hand padding)
- added requirement for user accounts, requires @nhs email for verification
- added links out to service desk and confluence etc.
- usable Django admin page, can monitor users + more scope for development
- removed all connections to Gemini database - should fix current unknown issue of having to run in debug mode
- removed redundant reanalysis and sample searching functionality
- better user messaging for handling errors - fixes #18 
- rewritten parts of primer designer for python3, improved formatting and remove unneeded code
- change DNAnexus-IGV to use dxpy instead of calling dx toolkit commands, faster and more reliable with no need for sourcing etc.
- IGV now fills full window instead of ~ 2/3
- IGV loads correctly with latest Firefox - fixes #32 
- handle providing development data through a dev project, with alerts before loading data - fixes #22 
- added ability to use hg19 or hg38 for loading IGV - fixes #24 
- added copy clipboard buttons to dx urls - fixes #25 
- searches matching multiple samples faster by rewriting calling dx generate_url after selecting sample

### CODE REVIEW

File: DNAnexus_to_igv/find_dx_002_bams.py [hunk 1]
```diff
@@ -37,23 +34,32 @@ def get_002_projects():
...
 
-    # get just the project id's from returned string
-    projects_002 = projects_002.replace("\n", " ").split(" ")
-    project_002_list = [x for x in projects_002 if x.startswith('project-')]
+    except Exception:
+        print(
+            "Failed getting id for project 003_200115_ga_igv_dev_data, "
+            "does it exist?"
+        )
+        pass
```
Comment (Yu-jinKim) [inline DNAnexus_to_igv/find_dx_002_bams.py:53]: why pass? no need for it

File: DNAnexus_to_igv/find_dx_002_bams.py [hunk 2]
```diff
@@ -81,129 +86,112 @@ def find_dx_bams(project_002_list):
...
+                    # check if bam is in CP tmp dir, pass if True
+                    continue
+
+                # index either can be .bam.bai or just .bai for old
+                # samples, get names for both to check
+                indexes = []
+                indexes.append(f"{bam_file}.bai")
+                indexes.append(f"{bam_file.strip('.bam')}.bai")
+
+                print(indexes)
```
Comment (Yu-jinKim) [inline DNAnexus_to_igv/find_dx_002_bams.py:139]: Still needed

File: DNAnexus_to_igv/views.py [hunk 3]
```diff
@@ -86,142 +122,201 @@ def nexus_search(request):
...
+                    find_dx_002_bams.py"""
+                ))
+                logging.error(IOe)
+
+                return render(
+                    request, 'DNAnexus_to_igv/nexus_search.html', context_dict
+                )
+
+            # select bams matching sample id, return original entry from
+            # JSON by mtahcing against upper name and search term
```
Comment (Yu-jinKim) [inline DNAnexus_to_igv/views.py:168]: typo

File: DNAnexus_to_igv/views.py [hunk 4]
```diff
@@ -86,142 +122,201 @@ def nexus_search(request):
...
+            sample_bams = [
+                v for k, v in json_bams.items() if sample_id.upper() in v[0]['bam_name'].upper()
```
Comment (Yu-jinKim) [inline DNAnexus_to_igv/views.py:170]: maybe use full name variable here to help comprehension
also i'd write:
```
[
    thing you extract
    for thing in things
    if condition
]
```

File: DNAnexus_to_igv/views.py [hunk 5]
```diff
@@ -86,142 +122,201 @@ def nexus_search(request):
...
                 # generate the urls
                 bam_url, idx_url = get_dx_urls(
-                                            sample_bams[0][0]["bam_file_id"],
-                                            sample_bams[0][0]["idx_file_id"]
-                                            )
-                
-                # add variables 
+                    request,
+                    sample_id,
+                    sample_bams[0][0]["bam_file_id"],
```
Comment (Yu-jinKim) [inline DNAnexus_to_igv/views.py:202]: from sample bams you take the only element in the list comprehension and you take the first element of that element which is a dict.
Maybe assign a variable like `bam = sample_bams[0]` or even `sample_bams[0][0]`

File: ga_core/settings.py [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+from .config import SECRET_KEY, PROD_HOST, DEBUG_HOST, GOOGLE_ANALYTICS,\
+    EMAIL_USER, EMAIL_PASSWORD
+
+# Build paths inside the project like this: BASE_DIR / 'subdir'.
+BASE_DIR = Path(__file__).resolve().parent.parent
+
+SECRET_KEY = SECRET_KEY
+
+# SECURITY WARNING: don't run with debug turned on in production!
+DEBUG = True
```
[the remark this answers is not in the document]
Comment (Yu-jinKim) [reply ga_core/settings.py:22]: You still haven't

File: primer_designer/views.py [hunk 7]
```diff
@@ -1,174 +1,128 @@
...
 
 
-def random_string(length=10):
-    """ Creates a random string 
+def random_string():
+    """
+    Creates a random string
 
     Args:
       length (int): length of string to generate, default 10
```
Comment (Yu-jinKim) [inline primer_designer/views.py:58]: there's no argument, remove docstring bit or add argument

File: setup_notes.md [hunk 8]
```diff
@@ -0,0 +1,30 @@
+## Notes for setting up / running Genetics Ark
+
+requirements:
+- install python packages from requirements.txt
+- add email account credentials to config.py to send account authorisation
+  emails from
+- setup send grid account and store API key linked to above email account in
+  config.py
+- recomended: create Django admin user (https://docs.djangoproject.com/en/3.1/intro/tutorial02/)
```
Comment (Yu-jinKim) [inline setup_notes.md:9]: typo

**Review by Yu-jinKim — COMMENTED** [review]
Some typos, some suggestions

File: DNAnexus_to_igv/views.py (hunk 5 above)
Comment (jethror1) [reply DNAnexus_to_igv/views.py:202]: that sensible, changed

File: DNAnexus_to_igv/views.py (hunk 4 above)
Comment (jethror1) [reply DNAnexus_to_igv/views.py:170]: also sensible, changed and added structure of output json to find bams script

File: DNAnexus_to_igv/find_dx_002_bams.py (hunk 1 above)
Comment (jethror1) [reply DNAnexus_to_igv/find_dx_002_bams.py:53]: true, it gone

File: setup_notes.md [hunk 9]
```diff
@@ -6,7 +6,7 @@ requirements:
   emails from
 - setup send grid account and store API key linked to above email account in
   config.py
-- recomended: create Django admin user (https://docs.djangoproject.com/en/3.1/intro/tutorial02/)
+- reccommended: create Django admin user (https://docs.djangoproject.com/en/3.1/intro/tutorial02/)
```
Comment (Yu-jinKim) [inline setup_notes.md:9]: jethro. come on

**Review by Yu-jinKim — COMMENTED** [review]
typo

**Review by Yu-jinKim — APPROVED** [review]

**Review by Yu-jinKim — APPROVED** [review]

**Review by mattgarner — APPROVED** [review]

**Pull request merged**